### PR TITLE
feat(storage): add StorageError and credentials validation

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

Implement the StorageError class and error messages map and ensure it aligns with the Auth category.

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by israx — COMMENTED** [review]
Looks good to me. No blocking but I'm wondering if we are going to expose the service errors as enums ?

File: packages/storage/src/errors/assertValidationErrors.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	StorageValidationErrorCode,
+	validationErrorMap,
+} from './types/validation';
+import { StorageError } from './StorageError';
+
+export function assertValidationError(
+	assertion: boolean,
+	name: StorageValidationErrorCode
+): asserts assertion {
+	const { message, recoverySuggestion } = validationErrorMap[name];
```
Comment (ashwinkumar6) [inline packages/storage/src/errors/assertValidationErrors.ts:14]: should we add `recoverySuggestion: ''` key to `validationErrorMap`

File: packages/storage/src/errors/StorageError.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+export enum StorageErrorTypes {
+	NO_CREDENTIALS = 'NO_CREDENTIALS',
+}
+
+const StorageErrorStrings: { [key in StorageErrorTypes]: ErrorParams } = {
+	[StorageErrorTypes.NO_CREDENTIALS]: {
+		name: 'NoCredentials',
+		message: 'No credentials',
+		recoverySuggestion: ``,
```
Comment (kvramyasri7) [inline packages/storage/src/errors/StorageError.ts:28]:
```suggestion
		recoverySuggestion: '',
```
But do we need to pass empty `recoverySuggestion` here?

**Review by kvramyasri7 — DISMISSED** [review]

File: packages/storage/src/errors/StorageError.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+		this.constructor = StorageError;
+		Object.setPrototypeOf(this, StorageError.prototype);
+	}
+}
+
+export enum StorageErrorTypes {
+	NO_CREDENTIALS = 'NO_CREDENTIALS',
+}
+
+const StorageErrorStrings: { [key in StorageErrorTypes]: ErrorParams } = {
```
[the remark this answers is not in the document]
Comment (AllanZhengYP) [reply packages/storage/src/errors/StorageError.ts:24]: Good catch, copy-pasta issue

**Review by ashwinkumar6 — APPROVED** [review]

**Review by kvramyasri7 — APPROVED** [review]
Looks good to me.

**Review by israx — APPROVED** [review]
LGTM!

**Pull request merged**